be/src/storage/persistent_index.h

```diff
@@ -116,63 +135,151 @@ class MutableIndex {
 
-    virtual Status append_wal(size_t n, const void* keys, const IndexValue* values,
-                              const std::vector<size_t>& idxes) = 0;
-
-    // get dump size of hashmap
-    virtual size_t dump_bound() = 0;
+    // load snapshot
+    bool load_snapshot(phmap::BinaryInputArchive& ar_in);
 
-    virtual bool dump(phmap::BinaryOutputArchive& ar_out) = 0;
+    // load according meta
+    Status load(const MutableIndexMetaPB& meta);
 
-    // [not thread-safe]
-    virtual size_t capacity() = 0;
+    size_t dump_bound();
 
-    virtual void reserve(size_t size) = 0;
+    bool dump(phmap::BinaryOutputArchive& ar_out);
 
-    virtual size_t memory_usage() = 0;
+    Status commit(MutableIndexMetaPB* meta, const EditVersion& version, const CommitType& type);
 
     // get all key-values pair references by shard, the result will remain valid until next modification
-    // |nshard|: number of shard
     // |num_entry|: number of entries expected, it should be:
     //                 the num of KV entries if without_null == false
     //                 the num of KV entries excluding nulls if without_null == true
     // |without_null|: whether to include null entries
-    // [not thread-safe]
-    virtual std::vector<std::vector<KVRef>> get_kv_refs_by_shard(size_t nshard, size_t num_entry,
-                                                                 bool without_null) const = 0;
+    std::vector<std::pair<uint32_t, std::vector<std::vector<KVRef>>>> get_kv_refs_by_shard(size_t num_entry,
+                                                                                           bool without_null);
 
-    // flush to immutable index
-    // [not thread-safe]
-    virtual Status flush_to_immutable_index(const std::string& dir, const EditVersion& version) const = 0;
+    std::vector<std::vector<size_t>> split_keys_by_shard(size_t nshard, const void* keys, size_t idx_begin,
+                                                         size_t idx_end);
+    std::vector<std::vector<size_t>> split_keys_by_shard(size_t nshard, const void* keys,
+                                                         const std::vector<size_t>& idxes);
 
-    static StatusOr<std::unique_ptr<MutableIndex>> create(size_t key_size, const std::string& path);
+    Status flush_to_immutable_index(const std::string& dir, const EditVersion& version);
+
+    // get the number of entries in the index (including NullIndexValue)
+    size_t size();
+
+    size_t capacity();
+
+    size_t memory_usage();
+
+private:
+    template <int N>
+    void _init_loop_helper();
+
+private:
+    constexpr static size_t kSliceMaxFixLength = 64;
+    uint32_t _fixed_key_size = -1;
+    uint64_t _offset = 0;
+    std::string _path;
+    std::unique_ptr<WritableFile> _index_file;
+    std::shared_ptr<FileSystem> _fs;
+    std::vector<std::unique_ptr<MutableIndex>> _shards;
+    std::vector<uint64_t> _page_sizes;
+    std::vector<size_t> _sizes;
```

Comment:
Maybe we only need one `_page_size` and `_size` but not an array. Just update the `_page_size` when append wal of shard.